File: 136. Single Number.java
class Solution {
    public int singleNumber(int[] nums) {
        // 关于XOR的阅读：https://stackoverflow.com/questions/1991380/what-does-the-operator-do-in-java
        // 对于本题：a^a^b^b^c^c^d^d^x = 0^0^0^0^x = x
        int ans =0;
        // 注意length后没有括号
        int len = nums.length;
        for(int i=0;i!=len;i++)
            ans ^= nums[i];
        return ans;
    }

    // method using set
    TODO


}

File: 136. Single Number.py
# space n 解法：hashmap/dict，或者hashset
class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        return 2*sum(set(nums))-sum(nums)


# XOR: 0^0 = 0, 0^1 = 1, 1^1 = 0
# 逻辑：偶数个数的数字会互相cancel，最后每位都是0，与single one做XOR操作，返回值为single one
class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        res = 0
        for num in nums:
            res ^= num
        return res


# XOR语法的其他写法：

# functools.reduce(fun,seq) applies a particular function passed in its argument to all of the list elements mentioned in the sequence passed along
class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        return reduce(lambda x, y: x ^ y, nums)
        
    def singleNumber(self, nums):
        return reduce(operator.xor, nums)

